### PR TITLE
Add DHCPv4 protocol to Packetbeat

### PR DESCRIPTION
**Pull request by andrewkroh**
Packetbeat will capture and index individual DHCP packets for IPv4.

Parsing is provided by https://github.com/insomniacslk/dhcp.

This expands the work done in elastic/beats#7359.

**Packetbeat DHCPv4 Overview Dashboard**

dashboard-packetbeat-dhcpv4

I created a sample watch to alert when new clients are detected on the network: https://gist.github.com/andrewkroh/b8fe93c0dead7eb963e37ac4ca1a332a

packetbeat-new-dhcp-client-detected

Note to self: https://play.golang.com/p/rHecAoaBmI9

### CODE REVIEW

**Comment by andrewkroh** [comment]
[I've been debating](https://github.com/elastic/beats/pull/7359#issuecomment-406308535) whether or not correlate request and responses.

"When did a client make a request and not get a response?" - This is a question you cannot answer without correlation. And we don't get a response time metric either.

I think response times could be measured for both discover/offer and request/ack [exchanges](https://tools.ietf.org/html/rfc2131#page-15).

- DHCPDISCOVER -> DHCPOFFER (multiple offers are possible)
  - Send one event for each offer so that we report the response time for each server.
- DHCPREQUEST -> DHCPACK/NAK (should be just one)
- DHCPDECLINE  (no measurement)
- DHCPRELEASE (no measurement)

But I think it could be a while before I personally have time to focus on this more. So perhaps we open an issue and track correlation as an enhancement and see if someone wants to take it on?

**Comment by andrewkroh** [comment]
jenkins please test this

**Comment by andrewkroh** [comment]
jenkins please test this

**Comment by andrewkroh** [comment]
I believe that the Windows build is failing because fields.yml is not generated during CI. https://github.com/elastic/beats/pull/7670 fixes the issue.

I don't understand how the system tests that validate fields names against fields.yml have been working or why it began failing in this PR.

**Comment by andrewkroh** [comment]
This is passing CI now, conflicts are resolved, and it's ready for review.

**Comment by andrewkroh** [comment]
> Do we flag new protocols as Experimental/Beta in Packetbeat?

We haven't done that before. But that did remind me to update the documentation to include DHCP.

**Comment by andrewkroh** [comment]
I'd like to do a squash this myself before merging to ensure that `Co-authored-by:` is added for @brianwaskiewicz since I used his pull-request as my starting point.

**Comment by andrewkroh** [comment]
I squashed it into two commits. One for vendor and one with our code.

Please use Rebase and Merge.

File: packetbeat/protos/dhcpv4/option_ip_addresses.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	code := dhcpv4.OptionCode(data[0])
+	length := int(data[1])
+	if length == 0 || length%4 != 0 {
+		return nil, fmt.Errorf("Invalid length: expected multiple of 4 larger than 4, got %v", length)
+	}
+	if len(data) < 2+length {
+		return nil, dhcpv4.ErrShortByteStream
+	}
+	servers := make([]net.IP, 0, length%4)
```
Comment (adriansr) [inline packetbeat/protos/dhcpv4/option_ip_addresses.go]: nit: I think you want `length / 4` here

Comment (andrewkroh) [reply packetbeat/protos/dhcpv4/option_ip_addresses.go]: Definitely. Good catch. I think I was looking at one of the parsers inside of the dhcp lib when I "wrote" that.

Will fix.

Comment (andrewkroh) [reply packetbeat/protos/dhcpv4/option_ip_addresses.go]: I fixed that line and squashed the commit.

**Comment by andrewkroh** [comment]
This needs back-porting to 6.x so that it gets included in the next minor (v6.5.0). 6.4.0 is already feature frozen.

**Pull request merged**